diesel_cli/src/main.rs

```diff
@@ -153,10 +153,20 @@ fn migration_version<'a>(matches: &'a ArgMatches) -> Box<Display + 'a> {
         .unwrap_or_else(|| Box::new(Utc::now().format(TIMESTAMP_FORMAT)))
 }
 
-fn migrations_dir(matches: &ArgMatches) -> PathBuf {
+fn migrations_dir_from_cli(matches: &ArgMatches) -> Option<PathBuf> {
     matches
         .value_of("MIGRATION_DIRECTORY")
         .map(PathBuf::from)
+        .or_else(|| {
```

Comment:
@weiznich - I'm still becoming familiar with rust idioms and clap! 😅. From what I understand, `Pathbuf::from` indicates that you expect the `MIGRATION_DIRECTORY` value will always convert without fail. Why do we need to chain `.or_else` and then match on the subcommand (name?, not sure which value of the tuple `1` represents)?

Reply:
> From what I understand, `Pathbuf::from` indicates that you expect the `MIGRATION_DIRECTORY` value will always convert without fail. 

As far as I understood the stdlib docs/code this conversion is guaranteed to succeed. If the user passes a nonsense path a `PathBuf` pointing to a non existing (the nonsense path) location is generated and the following operations executed by diesel will end with a error message indication a invalid/non existing path.

>Why do we need to chain `.or_else` and then match on the subcommand (name?, not sure which value of the tuple 1 represents)?

The `or_else` is needed to workaround some architectural decisions/limitations in clap. A clap app consists of several subcommands (migration, setup, print-schema , … for diesel_cli). Each subcommand could also consists of several subcommands (list, run, … for diesel migration). Arguments are defined in a subcommand and are valid for the defined subcommand and all child subcommands.  `ArgMatches` contains all arguments passed to a certain subcommand. Arguments are stored in that `ArgMatches` to which they are passed, not in which they are defined. So if someone is calling `diesel migration --migration_dir /some/dir list` the migration_dir argument is stored in the `ArgMatches` of migration. It is also possible to call `diesel migration list --migration_dir /some/dir` where migration_dir is stored in the `ArgMatches` of list. (There is a global option that allows a argument globally, but that does not solve this problem here). Therefore we need to traverse the results for each subcommand. This is done by recursively calling this method on each subcommand. (See the [clap docs](https://docs.rs/clap/2.27.1/clap/struct.ArgMatches.html#method.subcommand) for why the we access `.1` on the tuple)

Reply:
Awesome!!! 😄  Thanks so much! I'll finish review ASAP this week (unless someone else beats me to it).